features/net/network-socket/WiFiInterface.h

```diff
@@ -43,28 +45,78 @@ class WiFiInterface: public NetworkInterface
 public:
     /** Start the interface
      *
-     *  Attempts to connect to a WiFi network. If passphrase is invalid,
-     *  NSAPI_ERROR_AUTH_ERROR is returned.
+     *  Attempts to connect to a WiFi network.
      *
      *  @param ssid      Name of the network to connect to
      *  @param pass      Security passphrase to connect to the network
      *  @param security  Type of encryption for connection
-     *  @return          0 on success, negative error code on failure
+     *  @return          0 on success, or error code on failure
      */
-    virtual int connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
+    virtual nsapi_error_t connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
+
+    /** Start the interface
+     *
+     *  Attempts to connect to a WiFi network asynchronously, the call will return straight away. If the @a cb was NULL
+     *  you'll need to query @a get_state until it's in NSAPI_IF_STATE_CONNECTED state, otherwise the @a cb will be
+     *  called with connection results.
+     *
+     *  Note: @a ssid and @a pass must be kept until the connection is made, that is the callback has been called or the
+     *        state changed to @a NSAPI_IF_STATE_CONNECTED as they are passed by value.
+     *
+     *  @param ssid      Name of the network to connect to
+     *  @param pass      Security passphrase to connect to the network
+     *  @param security  Type of encryption for connection
+     *  @param cb        Function to be called when the connect finishes
+     *  @param data      Arbitrary user data to pass to @a cb function
+     */
+    virtual void connect_async(const char *ssid, const char *pass,nsapi_security_t security = NSAPI_SECURITY_NONE,
+                               wifi_connect_cb_t cb = NULL, void *data = NULL) = 0;
 
     /** Stop the interface
      *
-     *  @return          0 on success, negative error code on failure
+     *  @return          0 on success, or error code on failure
      */
-    virtual int disconnect() = 0;
+    virtual nsapi_error_t disconnect() = 0;
```

Comment:
Well I agree keeping them in sync would be good, but on the same note we sort of promised API for 5.2 for some partners to implement. If they do it once just so we can change the return type month later, they would have to go back to it and change it and also error handling code implemented for that 'new' API. Also I don't think there's much consistency required between interfaces on a technical level, as in it wouldn't cause anyone problems to have it different till we have time for the new PR. I would argue it will cause more confusion and work for apps to change the error checking code when we keep changing what the functions return.

I guess it's a call how much we don't want partners to go back and change things.